compiler/qsc_frontend/src/resolve.rs

```diff
@@ -131,7 +135,91 @@ impl Scope {
     }
 }
 
-struct GlobalScope {
+pub(super) type ScopeId = usize;
+
+#[derive(Debug, Clone, Default)]
+pub struct Locals {
+    // order is ascending by span (outermost -> innermost)
+    scopes: Vec<Scope>,
+}
+
+impl Locals {
+    pub(super) fn get_scopes<'a>(
+        &'a self,
+        scope_chain: &'a [ScopeId],
+    ) -> impl Iterator<Item = &Scope> + 'a {
+        // reverse to go from innermost -> outermost
+        scope_chain.iter().rev().map(|id| {
+            self.scopes
+                .get(*id)
+                .unwrap_or_else(|| panic!("scope with id {id:?} should exist"))
+        })
+    }
+
+    pub(super) fn push_scope(&mut self, s: Scope) -> ScopeId {
+        let id = self.scopes.len();
+        self.scopes.insert(id, s);
+        id
+    }
+
+    pub(super) fn get_scope_mut(&mut self, id: ScopeId) -> &mut Scope {
+        self.scopes
+            .get_mut(id)
+            .unwrap_or_else(|| panic!("scope with id {id:?} should exist"))
+    }
+
+    #[must_use]
+    pub fn get_all_at_offset(&self, offset: u32) -> Vec<Local> {
+        let mut vars = true;
+        let mut all_locals = Vec::new();
+        self.for_each_scope_at_offset(offset, |scope| {
+            // inner to outer
+            all_locals.extend(get_scope_locals(scope, offset, vars));
+
+            if scope.kind == ScopeKind::Callable {
+                // Since local callables are not closures, hide local variables in parent scopes.
+                vars = false;
+            }
+        });
+
+        // deduping by name will effectively make locals in a child scope
+        // shadow the locals in its parent scopes
+        all_locals.dedup_by(|a, b| a.name == b.name);
+
+        all_locals
+    }
+
+    fn for_each_scope_at_offset<F>(&self, offset: u32, mut f: F)
+    where
+        F: FnMut(&Scope),
+    {
+        // reverse to go from innermost -> outermost
+        self.scopes.iter().rev().for_each(move |scope| {
```

Comment:
I don't think so? What do you mean invalidate?

Reply:
Maybe I'm just misunderstanding a rust thing. I thought `move` would take each scope in `scopes` and move the data out of `scopes` such that it isn't in `scopes` anymore, or at least that it would be an error to reference `scopes` after this point.

Reply:
Ah I see. `move` only applies to the variables captured by the closure. In this case here, that's `f`. If we capture `f` by (immutable) reference, that makes rust unhappy as we're calling `f`, which mutates it. By using `move`, we make the closure _own_ `f` instead of borrowing it.
https://doc.rust-lang.org/std/keyword.move.html

As for `self.scopes`, there's nothing special there, we're just iterating over it using `iter()`. `scope` is a reference.

Reply:
Huh. I lied. While my general explanation still is (hopefully) correct, removing the `move` here _doesn't_ cause rust to complain, because `f` is already captured by value. I have no idea why I put this there then, must be from an earlier revision of the code!